Configurable fetch association behaviour [return array/relation]

daniellaniyo opened the pull request:
Fetch association to return array or relation

@dylanahsmith & @tjoyal for review/thoughts. Cc @camilo 

## Problem

There is an [open PR](https://github.com/Shopify/identity_cache/pull/224) that prevents chaining by having fetch association to return an array. However, it was never merged since it requires changes on the Shopify side. The idea of this PR to have a version where fetch association returns an array but also supports the current behaviour, i.e. returning a relation. This will allow us to replace these relations dependencies gradually on the Shopify side while having the correct IDC default behaviour.

Once everything is cleaned up, we will plan a major release that only returns arrays.

## Solution

By default, return an Array from the fetch association method rather than a relation to prevent chaining.  Any time we need to support the old behaviour, use `IdentityCache.allow_fetch_to_return_relation` around the block. 

A first test was done on [this shopify branch](https://github.com/Shopify/shopify/compare/bump-IDC) to confirm that [tests pass](https://buildkite.com/shopify/shopify-branches/builds/112661) with these changes.  `IdentityCache.allow_fetch_to_return_relation`  will be used around tests that still depend on relations returned by fetch_association.

dylanahsmith commented on CHANGELOG.md:
```diff
@@ -4,6 +4,10 @@
 
 - Stop sharing the same attributes hash between the fetched record and the memoized cache, which could interfere with dirty tracking
```
wouldn't this be part of the release too?

daniellaniyo replied:
Yes, it will! I have updated the changelog in my last commit.

dylanahsmith commented on CHANGELOG.md:
```diff
@@ -4,6 +4,10 @@
 
+#### 0.3.2
+
+- Fetch association returns relation or array depending on the configuration
```
This isn't clear about how it is configured by default.  It should behave the same way by default until 0.4 is released.

dylanahsmith replied:
We might also want to mention that it was only returning a relation for cache_has_many fetch association methods

dylanahsmith commented on lib/identity_cache.rb:
```diff
@@ -40,6 +40,9 @@ class << self
     self.cache_namespace = "IDC:#{CACHE_VERSION}:".freeze
 
+    mattr_accessor :fetch_returns_relation
+    self.fetch_returns_relation = false
```
This should be true to avoid having this release be a breaking change

daniellaniyo replied:
Done! Now, set to false. I have added an `allow_fetch_to_return_array` method to go along with this change.

dylanahsmith commented on lib/identity_cache/configuration_dsl.rb:
```diff
@@ -236,7 +236,11 @@ def #{options[:cached_accessor_name]}
+              if IdentityCache.fetch_returns_relation
+                #{association}
+              else
+                association(:#{association}).load_target
```
I think `#{association}.to_a` would be preferred since that is part of the documented active record API.

dylanahsmith commented on lib/identity_cache/query_api.rb:
```diff
@@ -398,18 +404,30 @@ def prefetch_one_association(association, records)
+          if instance_variable_defined?(ivar_full_name)
+            instance_variable_get(ivar_full_name)
+          else
+            instance_variable_set(ivar_full_name, association(association_name).load_target)
```
I think it will be easier to reason about the code if we always store the samr thing in this instance variable and just call to_a on the result if IdentityCache.fetch_returns_relation returns false

camilo commented on lib/identity_cache/query_api.rb:
```diff
@@ -398,18 +404,30 @@ def prefetch_one_association(association, records)
+          assoc.is_a?(ActiveRecord::Associations::CollectionAssociation) ? assoc.reader : assoc
         else
-          instance_variable_set(ivar_full_name, send(association_name))
+          send(association_name.to_sym)
```
`public_send` ?

daniellaniyo commented:
Thanks for your feedback! Please check [latest commit](https://github.com/Shopify/identity_cache/pull/276/commits/0a16f2cb38bdda9e9bfae45b9cebd01a30fc10be?diff=split) for changes.

camilo commented:
LGTM wait for @dylanahsmith tho

dylanahsmith commented on lib/identity_cache.rb:
```diff
@@ -131,6 +134,18 @@ def fetch_multi(*keys)
+    def allow_fetch_to_return_relation
+      self.fetch_returns_relation = true
+      yield
+      self.fetch_returns_relation = false
```
To avoid leaky behaviour, we usually store the old value in a local variable then use an ensure block to set it back to that old value.

You could also merge this function with the following one by just parameterizing the value.  E.g. `def with_ fetch_returns_relation(value=true)`

dylanahsmith commented on lib/identity_cache/query_api.rb:
```diff
@@ -156,7 +156,13 @@ def set_embedded_association(record, association_name, coder_or_array) #:nodoc:
+          embedded_variable.is_a?(Array)
+        end
+
+        if cached_association
```
What you really care about is if embedded_variable is array like, which can be checked for with just `if embedded_variable.respond_to?(:to_ary)`, then you can use `embedded_variable.to_ary.map {`... on the following line

dylanahsmith commented on lib/identity_cache/query_api.rb:
```diff
@@ -400,16 +406,18 @@ def prefetch_one_association(association, records)
           instance_variable_set(ivar_full_name, send(association_name))
         end
-
-        assoc.is_a?(ActiveRecord::Associations::CollectionAssociation) ? assoc.reader : assoc
```
This is a weird case.  It doesn't look like `send(association_name)` would return a ActiveRecord::Associations::CollectionAssociation.  I think this is working around an inconsistency that shouldn't be there.  I'm going to open up a pull request to fix this, since it looks like it is causing some confusion.

dylanahsmith replied:
See https://github.com/Shopify/identity_cache/pull/277

dylanahsmith commented on lib/identity_cache/query_api.rb:
```diff
@@ -400,16 +406,18 @@ def prefetch_one_association(association, records)
+        if IdentityCache.fetch_returns_relation
+          assoc.is_a?(ActiveRecord::Associations::CollectionAssociation) ? assoc.reader : assoc
+        else
+          assoc.to_a
```
This method is used for cache_has_one's fetch association method as well, so we can't always call `.to_a` if `IdentityCache.fetch_returns_relation` is false

daniellaniyo replied:
Would using `load_target` be a good idea here for cases where `IdentityCache.fetch_returns_relation` is false?

dylanahsmith replied:
No, I think that should be handled separately.  I think this is what we need here

```ruby
assoc = assoc.reader if assoc.is_a?(ActiveRecord::Associations::CollectionAssociation)
assoc = assoc.to_ary if assoc.respond_to?(:to_ary) && !IdentityCache.fetch_returns_relation
assoc
```

dylanahsmith replied:
Actually, the `assoc = assoc.to_ary if assoc.respond_to?(:to_ary) && !IdentityCache.fetch_returns_relation` line should probably happen outside of `if IdentityCache.should_use_cache?`

dylanahsmith commented on test/denormalized_has_many_test.rb:
```diff
@@ -12,7 +12,23 @@ def setup
   end
 
+  def test_uncached_record_from_the_db_will_use_normal_association_as_ary
+    # Test configuration where fetch association returns an array
```
Try to make the code more readable by using more appropriate variable names before adding comments.  For instance, the test name doesn't mention the configuration or `allow_fetch_to_return_array`, but it looks like that is what you are actually testing.

dylanahsmith commented on test/denormalized_has_many_test.rb:
```diff
@@ -12,7 +12,23 @@ def setup
+      expected = @record.associated_records.to_a
+      record_from_db = Item.find(@record.id)
+      assert_queries(1) do
+        assert_memcache_operations(0) do
```
I think you are trying to test too many things here.  Try to keep the test cases more focused.  I don't think you need to test the number of queries in this case.

daniellaniyo replied:
I could surely focus on the class of the fetch result.

I had the number of queries checked to make sure we had the right behaviour for both cached & uncached records; and ensure tests in the previous [PR](https://github.com/Shopify/identity_cache/pull/224/files) were still passing.

dylanahsmith commented on test/denormalized_has_many_test.rb:
```diff
@@ -12,7 +12,23 @@ def setup
+      record_from_db = Item.find(@record.id)
+      assert_queries(1) do
+        assert_memcache_operations(0) do
+          assert_equal expected, record_from_db.fetch_associated_records
```
It looks like a collection proxy implements `==` to compare the loaded target to the other object, so I don't think this will actually catch a regression.  I think the assertion you actually want is `assert_equal Array, @record.fetch_associated_records.class`

dylanahsmith replied:
Actually, it looks like the denormalized and normalized test_fetch_association_does_not_allow_chaining_with_ary tests already provide regression tests.  What else are you trying to test with these other tests?

dylanahsmith commented on test/denormalized_has_many_test.rb:
```diff
@@ -12,7 +12,23 @@ def setup
+
   def test_uncached_record_from_the_db_will_use_normal_association
+    # Test configuration where fetch association returns a relation
+    assert_equal IdentityCache.fetch_returns_relation, true
```
I don't think this is relevant for this test.

daniellaniyo commented:
@dylanahsmith Could you please check out [the latest commit](https://github.com/Shopify/identity_cache/pull/276/commits/b09c5f6798189f641b823d890eca966cc7609d4a)? Thoughts?

daniellaniyo commented:
Rebased to grab [277](https://github.com/Shopify/identity_cache/pull/277) changes and squashed previous commits. @dylanahsmith Could you please check out [the latest commit](https://github.com/Shopify/identity_cache/pull/276/files?diff=split)? Thoughts?

dylanahsmith commented on lib/identity_cache.rb:
```diff
@@ -131,6 +134,14 @@ def fetch_multi(*keys)
+      previous_value = self.fetch_returns_relation
+      self.fetch_returns_relation = value
+      yield
+      ensure
```
This should be indented only as much as `def`.  The following line also needs less indentation

daniellaniyo replied:
Aaaahh my bad...derp

dylanahsmith commented on lib/identity_cache/query_api.rb:
```diff
@@ -401,13 +401,15 @@ def fetch_recursively_cached_association(ivar_name, association_name) # :nodoc:
       if IdentityCache.should_use_cache?
```
move `assoc = ` to this line and 

```ruby
        assoc = assoc.to_ary if assoc.respond_to?(:to_ary) && !IdentityCache.fetch_returns_relation
        assoc
```

after this condition.  That way the else block can be changed back to `send(association_name.to_sym)` and the conditional array coercion only needs to happen in one place in this method

dylanahsmith commented on CHANGELOG.md:
```diff
@@ -1,7 +1,8 @@
-#### Unreleased
+#### 0.3.2
 
+- Fetch association returns relation or array depending on the configuration. It was only returning a relation for cache_has_many fetch associtions methods.
```
typo: `associtions` -> `association`

Also, I think we should start referencing pull requests or issues in changelog entries by adding `(#276)` to the end to entry

daniellaniyo replied:
Done. I've added the PR numbers for the 0.3.2 entries. See commit below.

daniellaniyo commented:
@dylanahsmith Thanks for the review! Please see the latest commit [here](https://github.com/Shopify/identity_cache/pull/276/commits/8aeb2e9307edb8a690ce620efb5b75515914ec70).

dylanahsmith commented:
:+1:

daniellaniyo commented:
Great! Squashing and merging.
